control-plane/csi-driver/src/bin/node/node.rs

```diff
@@ -432,9 +437,117 @@ impl node_server::Node for Node {
+        // amongst the various error types in this crate.
+        type CsiAccessTypeError = String;
+
+        // An enum to assess the state of volume when expecting it to be a filesystem type.
+        enum CsiAccessType {
+            SupportedFilesystem(FileSystem),
+            UnsupportedFilesystem(CsiAccessTypeError),
+            NotAFilesystem,
+            Unknown(CsiAccessTypeError),
+        }
+
+        // From VolumeCapability translate volume access-type into one of few known states.
+        impl From<&Option<VolumeCapability>> for CsiAccessType {
+            fn from(volume_capability: &Option<VolumeCapability>) -> Self {
+                match get_access_type(volume_capability) {
+                    Ok(access_type) => match access_type {
+                        AccessType::Mount(mv) => match mv.fs_type.to_lowercase().as_str() {
+                            "ext4" => Self::SupportedFilesystem(FileSystem::from(Fs::Ext4)),
+                            "xfs" => Self::SupportedFilesystem(FileSystem::from(Fs::Xfs)),
+                            "btrfs" => Self::SupportedFilesystem(FileSystem::from(Fs::Btrfs)),
+                            alien_fs => Self::UnsupportedFilesystem(format!(
+                                "'{alien_fs}' is not a supported filesystem"
+                            )),
+                        },
+                        AccessType::Block(_) => Self::NotAFilesystem,
+                    },
+                    Err(error) => {
+                        Self::Unknown(format!("couldn't determine CSI AccessType type: {}", error))
+                    }
+                }
+            }
+        }
+        //===============================CsiAccessType=============================================
+
+        let request = request.into_inner();
+
+        let vol_uuid = Uuid::parse_str(request.volume_id.as_str()).map_err(|error| {
+            Status::invalid_argument(format!(
+                "Malformed volume UUID '{}': {error}",
+                request.volume_id
+            ))
+        })?;
+
+        if request.volume_path.is_empty() {
+            return Err(Status::invalid_argument("'volume_path' is empty"));
+        }
+
+        let _guard = VolumeOpGuard::new(vol_uuid)?;
+
+        let dev_name = Device::lookup(&vol_uuid)
+            .await
+            .map_err(|error| {
+                Status::internal(format!(
+                    "device lookup error for volume '{vol_uuid}': {error}"
+                ))
+            })?
+            .ok_or(Status::invalid_argument(format!(
+                "failed to find a device for volume {vol_uuid}"
+            )))?
+            .devname();
+        // Get device size.
+        // The underlying block device should already have been expanded to the
+        // required size as a part of the ControllerExpandVolume call.
+        let capacity_bytes = get_size_from_dev_name(dev_name.as_str()).map_err(|error| {
```

Comment:
This is a good idea. I think to perform this check correctly, we'd have to get the usable blockdevice capacity from the total blockdevice capacity. Otherwise a comparision against this size value would be hacky. Wdyt @tiagolobocastro @dsharma-dc ?

Reply:
Yes, the usable capacity exported by nexus is typically bit lesser than spec after accounting various things, but `get_size_from_dev_name` will return the usable amount only I guess so that should be ok. However, since we don't specify size in `resize2fs` or `xfs_grow` commands, it'll figure out the usable size automatically. Comparing against what we expect may not be very fruitful here.

Reply:
While this plugin method doesn't handle the expansion for the block layer, the CSI spec requires that we only return OK iff the allocated capacity for the device is equal to or more than the required capacity. This implies that the less-than case is an error case. 
In principle, the controller plugin should return that kind of an error, but that would be quite difficult to reach out to linux API for for nodes from the controller plugin. However, the node plugin can do that.

Reply:
Oh, I think I get why the /sys/block size is completely usable now.

Reply:
@dsharma-dc  Why is the capacity exported by the nexus less than spec?
@niladrih can't we just test capacity_bytes against the request?

Reply:
Oh that's right, we probably should create replicas with extra capacity to account for this...
But then it seems like we need to check here against size - 5MiB @niladrih ?

Reply:
A bit confusing part here is that the -5MiB is from the actual replica size, and not requested size. Example below:

- Create 105MiB replica and a nexus with it -> spdk replica bdev size is actually created 108MiB, and nexus bdev is 105MiB. Nexus size visible to host is 103MiB(108 - 5)
- Resize replica and nexus to 150MiB -> spdk replica bdev becomes 152MiB, nexus bdev is 150MiB. Nexus size visible to host is 147MiB (152 - 5)

so a check here like below seems ok perhaps?
`(capacity_bytes > requested_size - 5MiB) => OK`

Reply:
Is this in a const somewhere I can use, or do you want me to use the 5 MiB raw?
@tiagolobocastro @dsharma-dc ?

Reply:
Also, it seems like the nexus size is 3 away from REST Volume resource size? Is the Volume resource size the same as the required_bytes from CreateVolumeRequest?

Reply:
> Also, it seems like the nexus size is 3 MiB away from REST Volume resource size? Is the Volume resource size the same as the required_bytes from CreateVolumeRequest?

There is no const to use as such, you can define one and use perhaps. The volume resource size will be same as REST volume resource size, which will same in resource spec. The extras that we we see in bdev are done at spdk for book-keeping.

Reply:
Adjusted the required_bytes by 5MiB.